script/pgsqlms

```diff
@@ -74,6 +74,186 @@ my $ATTRD_PRIV    = "$HA_SBIN_DIR/attrd_updater --private --lifetime reboot";
 # Global vars
 my $nodename;
 my $exit_code = 0;
+# numeric pgsql versions
+my $PGVERNUM;
+my $PGVER_10  = 100000;
+
+# Run a query using psql.
+#
+# This function returns an array with psql return code as first element and
+# the result as second one.
+#
+sub _query {
+    my $query        = shift;
+    my $res          = shift;
+    my $connstr      = "dbname=postgres";
+    my $RS           = chr(30); # ASCII RS  (record separator)
+    my $FS           = chr(3);  # ASCII ETX (end of text)
+    my $postgres_uid = getpwnam( $system_user );
+    my $oldeuid      = $>;
+    my $tmpfile;
+    my @res;
+    my $ans;
+    my $pid;
+    my $rc;
+
+    unless ( defined $res and defined $query and $query ne '' ) {
+        ocf_log( 'debug', '_query: wrong parameters!' );
+        return -1;
+    }
+
+    unless ( $tmpfile = File::Temp->new(
+            TEMPLATE => 'pgsqlms-XXXXXXXX',
+            DIR      => $HA_RSCTMP
+        ) )
+    {
+        ocf_exit_reason( 'Could not create or write in a temp file' );
+        exit $OCF_ERR_INSTALLED;
+    }
+
+    print $tmpfile $query;
+    chmod 0644, $tmpfile;
+
+    ocf_log( 'debug', '_query: %s', $query );
+
+    # Change the effective user to the given system_user so after forking
+    # the given uid to the process should allow psql to connect w/o password
+    $> = $postgres_uid;
+
+    # Forking + piping
+    $pid = open(my $KID, "-|");
+
+    if ( $pid == 0 ) { # child
+        exec $PGPSQL, '--set', 'ON_ERROR_STOP=1', '-qXAtf', $tmpfile,
+            '-R', $RS, '-F', $FS, '--port', $pgport, '--host', $pghost,
+            $connstr;
+    }
+
+    # parent
+    $> = $oldeuid;
+
+    {
+        local $/;
+        $ans = <$KID>;
+    }
+
+    close $KID;
+    $rc = $? >> 8;
+
+    ocf_log( 'debug', '_query: psql return code: %d', $rc );
+
+    if ( defined $ans ) {
+        chop $ans;
+
+        push @{ $res }, [ split(chr(3) => $_, -1) ]
+            foreach split (chr(30) => $ans, -1);
+
+        ocf_log( 'debug', '_query: @res: %s',
+            Data::Dumper->new( [ $res ] )->Terse(1)->Dump );
+    }
+
+    # Possible return codes:
+    #  -1: wrong parameters
+    #   0: OK
+    #   1: failed to get resources (memory, missing file, ...)
+    #   2: unable to connect
+    #   3: query failed
+    return $rc;
+}
+
+# Get the last received location on a standby
+# if the first argument is true, returns the value as decimal
+# if the first argument is false, returns the value as LSN
+# Returns undef if query failed
+sub _get_last_received_lsn {
```

Comment:
I created this sub tho "hide" the mess with pg10 compat around the query from the callers.